desc/protoparse/lexer.go

```diff
@@ -761,8 +753,13 @@ func (l *protoLex) skipToEndOfBlockComment() (string, bool) {
 	}
 }
 
-func (l *protoLex) Error(s string) {
-	if l.err == nil {
-		l.err = ErrorWithSourcePos{Underlying: errors.New(s), Pos: l.prevSym.start()}
+func (l *protoLex) addError(err error) {
+	if _, ok := err.(ErrorWithPos); !ok {
+		err = ErrorWithSourcePos{Pos: l.prev(), Underlying: err}
 	}
+	l.errs = append(l.errs, err)
+}
+
+func (l *protoLex) Error(s string) {
```

Comment:
This really bugs me - see commentary I put at top-level

Reply:
This is going to result in errors of type `*errors.errorString` https://golang.org/src/errors/errors.go?s=293:320#L1 via `addError` above, not of type `ErrorWithPos`

Reply:
As written, it will not. See the implementation of `addError`, which conditionally wraps it in an `ErrorWithSourcePos`.

The problem I think I see is the opposite: that system errors will also be incorrectly wrapped. So I'll fix that in the next push.

Reply:
Ah I got the `ok!` vs `ok` mixed up.

I think what would be great after this PR is for `ParseFiles*` to have a clear contract as to what error types are returned in what situations. Even better would be for there to be new methods `ParseFilesV2*` or some differentiator other than V2 that return `(*ParseFilesOutput, error)`, where `error` is returned if there is a system error, and `ParseFilesOutput` is:

```go
type ParseFilesOutput struct {
  FileDescriptors []*desc.FileDescriptor
  ParseErrors []ErrorWithPos
```

Where only one of the values is set.